### ISSUE TITLE
set_task_pid: deterministic "host pid is error" when CUDA_VISIBLE_DEVICES does not start with the pod's device 0 (probe context and poll target mismatch), leading to spurious Device 0 OOM

### ISSUE THREAD

**Issue by ihnokim** [Project-HAMi/HAMi-core#225]
## Summary

In a container with multiple vGPUs, `set_task_pid()` fails deterministically ("host pid is error!") for any process whose `CUDA_VISIBLE_DEVICES` does not **start with** the container's device 0. The failure is not a race: it reproduces 100% with a single process and no concurrency. After the failure, the process's memory usage is merged into the wrong accounting slot, so allocations that are legal per device get summed against one device's limit and trip a spurious `Device 0 OOM`.

This looks like the root cause behind reports such as Project-HAMi/HAMi#1455 (same log signature during multi-GPU training, closed as stale without analysis).

## Environment

- HAMi v2.7.1 (bundled libvgpu), also traced against HAMi-core `main` sources
- Kubernetes + hami-scheduler, pod with `nvidia.com/gpu: 2`, `nvidia.com/gpumem: 5800` (per device)
- NVIDIA driver 580.82.07, RTX 2080 Ti x2 (originally observed on A100-80GB x8, driver 580.105, verl FSDP world=8)
- PyTorch 2.x (`vllm/vllm-openai:v0.10.2` image), CUDA 12.x

## Reproduction

Inside a single pod that was allocated 2 vGPUs, run one Python process per case (no concurrency at all), each allocating 4096 MiB in 256 MiB steps (well under the 5800 MiB per-device cap):

```bash
CUDA_VISIBLE_DEVICES=<case> python3 -c '
import sys, torch
# optional: torch.cuda.set_device(1)
torch.cuda.init()
bufs = [torch.empty(256 << 20, dtype=torch.uint8, device="cuda") for _ in range(16)]
'
```

Results (3/3 identical across repetitions):

| case | CUDA_VISIBLE_DEVICES | context created on | result |
|---|---|---|---|
| A | `0` | pod dev 0 | ok |
| B | `1` | pod dev 1 | **host pid is error** |
| C | `1,0` | pod dev 1 (default) | **host pid is error** |
| D | `1,0` + `set_device(1)` | pod dev 0 | **host pid is error** |
| E | `0,1` + `set_device(1)` | pod dev 1 | ok |
| F | `0,1` | pod dev 0 | ok |

So the failure condition is exactly: **`CUDA_VISIBLE_DEVICES[0]` != the pod's first device**. Where the user code later places its context is irrelevant (C vs D vs E), because the probe happens during library initialization.

With two workers (CVD=`0` and CVD=`1`, e.g. Ray actors in one pod, or any launcher that narrows CVD per rank), the CVD=`1` worker fails host-pid detection every time, and its allocations are accounted together with the other worker's, producing `Device 0 OOM <sum>/<cap>` while each device is individually within its cap. Observed signatures: `Device 0 OOM 6337593344 / 6081740800` (2080 Ti), `88.79GB / 80GB` (A100 x8).

## Root cause (from source)

In `src/utils.c`, `set_task_pid()`:

1. The before/after process snapshots are taken from the **first CVD-visible device in NVML (physical) enumeration order** — the loop does an unconditional `break` after the first device where `nvml_to_cuda_map(i) >= 0`. For CVD=`1,0` or `0,1` that is the pod's physical device 0.
2. The probe context, however, is created with `cuDevicePrimaryCtxRetain(&pctx, 0)` — CUDA device 0, i.e. **whatever CVD lists first**.

When CVD=`1,0`, the poll watches physical device 0 while the probe context appears on physical device 1, so `getextrapid()` never sees a new PID and returns 0 -> `LOG_ERROR("host pid is error!")`.

When CVD=`1`, physical device 0 is not visible at all, `nvml_to_cuda_map(0)` returns -1 for every polled index that could match, no snapshot is taken (`previous == running == 0`), and `getextrapid()` again returns 0. Same error, different path.

All six cases above, plus the original multi-worker OOM, are consistent with this model.

## Suggested fix

Create the probe context on the same device that is being polled, e.g. retain the primary context for the `cudaDev` found by the map inside the loop instead of hardcoding device 0:

```c
cudaDev = nvml_to_cuda_map(i);
if (cudaDev < 0) continue;
/* poll nvml index i ... */
CHECK_CU_RESULT(cuDevicePrimaryCtxRetain(&pctx, cudaDev));
```

(or equivalently, poll the NVML device that corresponds to CUDA device 0). Either direction removes the mismatch and also fixes the CVD=`1` path, since the polled device and the probe device are then guaranteed to be the same.

## Workarounds we validated

- Keep the pod's first device first in CVD (e.g. `0,1` is safe, `1,0` is not), or
- unset CUDA_VISIBLE_DEVICES in the container and select devices with `torch.cuda.set_device()` per process, or
- use a 1-GPU-per-pod layout so every process's CVD starts with that pod's device 0.

Happy to provide the full reproduction manifests (Kubernetes pod specs) if useful.

**Comment by maverick-woo** [comment]
Thanks for the excellent and thorough analysis @ihnokim, this is really well done.
  
  I've confirmed the root cause in src/utils.c:set_task_pid():

  - The before/after process snapshots (lines 114–149) poll the first NVML device with a valid CUDA mapping (the break after the first nvml_to_cuda_map(i) >= 0 hit)
  - But cuDevicePrimaryCtxRetain(&pctx, 0) on line 133 always creates the probe context on CUDA device 0, which is CUDA_VISIBLE_DEVICES[0] — a different physical device whenever CVD doesn't start with the pod's
  device 0

  When they mismatch, getextrapid() finds no new PID and we get the spurious "host pid is error!" → accounting lands on the wrong device → fake OOM.

  Your suggested fix looks correct. I think the cleanest approach is to capture the cudaDev from the first loop and use it for both the retain and release calls, something like:

  int probeDev = 0;
  for (i = 0; i < nvmlCounts; i++) {
      cudaDev = nvml_to_cuda_map(i);
      if (cudaDev < 0) continue;
      // ... snapshot ...
      probeDev = cudaDev;
      break;
  }
  // ...
  CHECK_CU_RESULT(cuDevicePrimaryCtxRetain(&pctx, probeDev));
  // ... second loop ...
  CHECK_CU_RESULT(cuDevicePrimaryCtxRelease(probeDev));

**Comment by maverick-woo** [comment]
fixed in https://github.com/Project-HAMi/HAMi-core/pull/230

**Comment by ihnokim** [comment]
Thanks for the quick confirmation and the fix, really appreciate the fast turnaround. The approach in #230 of reusing the `cudaDev` found in the enumeration loop for both the retain and release calls is exactly what I had in mind.

I still have the 2x RTX 2080 Ti repro environment available, so I'd be happy to run the case table above against the patched build if that would help with verification.

**Comment by maverick-woo** [comment]
@ihnokim Thanks for offering to test! We've already verified the fix on an 8x RTX 3090 with all four CVD combinations (0, 1, 0,1, 1,0) — all passed with no "host pid is error!". The fix is in #230 pending review.
